File: nd4j/nd4j-backends/nd4j-api-parent/nd4j-api/src/main/java/org/nd4j/linalg/api/ops/impl/shape/DiagPart.java
/*******************************************************************************
 * Copyright (c) 2015-2018 Skymind, Inc.
 *
 * This program and the accompanying materials are made available under the
 * terms of the Apache License, Version 2.0 which is available at
 * https://www.apache.org/licenses/LICENSE-2.0.
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS, WITHOUT
 * WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied. See the
 * License for the specific language governing permissions and limitations
 * under the License.
 *
 * SPDX-License-Identifier: Apache-2.0
 ******************************************************************************/

package org.nd4j.linalg.api.ops.impl.shape;

import onnx.OnnxProto3;
import org.nd4j.autodiff.samediff.SDVariable;
import org.nd4j.autodiff.samediff.SameDiff;
import org.nd4j.base.Preconditions;
import org.nd4j.linalg.api.ndarray.INDArray;
import org.nd4j.linalg.api.ops.DynamicCustomOp;
import org.tensorflow.framework.AttrValue;
import org.tensorflow.framework.GraphDef;
import org.tensorflow.framework.NodeDef;

import java.util.Arrays;
import java.util.Collections;
import java.util.List;
import java.util.Map;


/**
 * Return the diagonal part of a tensor. The input tensor has to
 * have dimensions [d1,..., dk, d1,..., dk], so that the diagonal
 * blocks have shape [d1,..., dk].
 * <p>
 * A simple special case of this is returning the diagonal of a
 * matrix as vector.
 *
 * @author Max Pumperla
 */
public class DiagPart extends DynamicCustomOp {

    public DiagPart() {
    }

    public DiagPart(SameDiff sameDiff, SDVariable[] args, boolean inPlace) {
        super(null, sameDiff, args, inPlace);
    }

    public DiagPart(INDArray in, INDArray out){
        super(null, in, out, null, null);
    }

    @Override
    public List<SDVariable> doDiff(List<SDVariable> i_v) {
        SDVariable grad = i_v.get(0);
        SDVariable ret = sameDiff.diag(grad);
        return Collections.singletonList(ret);
    }

    @Override
    public String opName() {
        return "diag_part";
    }


    @Override
    public String tensorflowName() {
        return "DiagPart";
    }

    @Override
    public void initFromTensorFlow(NodeDef nodeDef, SameDiff initWith, Map<String, AttrValue> attributesForNode, GraphDef graph) {
        super.initFromTensorFlow(nodeDef, initWith, attributesForNode, graph);
    }

    @Override
    public void initFromOnnx(OnnxProto3.NodeProto node, SameDiff initWith, Map<String, OnnxProto3.AttributeProto> attributesForNode, OnnxProto3.GraphProto graph) {
        super.initFromOnnx(node, initWith, attributesForNode, graph);
    }

    @Override
    public List<org.nd4j.linalg.api.buffer.DataType> calculateOutputDataTypes(List<org.nd4j.linalg.api.buffer.DataType> dataTypes){
        Preconditions.checkState(dataTypes.size() == 1, "Expected list with exactly 1 datatype, got %s", dataTypes);
        //Output type is same as input type
        return dataTypes;
    }

}

File: libnd4j/include/ops/declarable/generic/parity_ops/diagPart.cpp
/*******************************************************************************
 * Copyright (c) 2015-2018 Skymind, Inc.
 *
 * This program and the accompanying materials are made available under the
 * terms of the Apache License, Version 2.0 which is available at
 * https://www.apache.org/licenses/LICENSE-2.0.
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS, WITHOUT
 * WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied. See the
 * License for the specific language governing permissions and limitations
 * under the License.
 *
 * SPDX-License-Identifier: Apache-2.0
 ******************************************************************************/

//
// @author Yurii Shyrma (iuriish@yahoo.com), created on 06.12.2017.
//

#include <op_boilerplate.h>
#if NOT_EXCLUDED(OP_diag_part)

#include <ops/declarable/CustomOperations.h>


namespace nd4j {
namespace ops  {
		
		CUSTOM_OP_IMPL(diag_part, 1, 1, false, 0, 0) {
			auto input  = INPUT_VARIABLE(0);
    		auto output = OUTPUT_VARIABLE(0);

    		const int inRank = input->rankOf();
    
    		// input validation
    		REQUIRE_TRUE(inRank == 2 ||  inRank == 4 || inRank == 6, 0, "DIAG_PART op: input array must have rank among following three possible values: 2, 4, 6, but got %i instead !", inRank);
    		for(int i = 0; i < inRank-1; ++i)
    			REQUIRE_TRUE(input->sizeAt(i) == input->sizeAt(i+1), 0, "DIAG_PART op: wrong shape of input array %s ! All dimensions must be equal !", ShapeUtils::shapeAsString(input).c_str());

    		const int outLen = output->lengthOf();
    		const int inLen  = input->lengthOf();

    		int i(0), j(0);
    		// FIXME: must be moved to helper
    		while(j < outLen) {
    			//(*output)(j) = (*input)(i);
    			//i += outLen+1;
    			//++j;
    			throw std::runtime_error("Not implemented yet");
    		}
    
		    return Status::OK();
		}
		DECLARE_SYN(DiagPart, diag_part);


		DECLARE_SHAPE_FN(diag_part) {
    		auto inputShapeInfo = inputShape->at(0);

    		const int inRank = inputShapeInfo[0];

    		// input validation
    		REQUIRE_TRUE(inRank == 2 ||  inRank == 4 || inRank == 6, 0, "DIAG_PART op: input array must have rank among following three possible values: 2, 4, 6, but got %i instead !", inRank);
    		for(int i = 1; i < inRank; ++i)
    			REQUIRE_TRUE(inputShapeInfo[i] == inputShapeInfo[i+1], 0, "DIAG_PART op: wrong shape of input array %s ! All dimensions must be equal !", ShapeUtils::shapeAsString(inputShapeInfo).c_str());

    		Nd4jLong* outShapeInfo = nullptr;

			int outRank = inRank/2;

			ALLOCATE(outShapeInfo, block.getWorkspace(), shape::shapeInfoLength(outRank), Nd4jLong);
	
			outShapeInfo[0] = outRank;
			for(int i = 1; i <= outRank; ++i)
				outShapeInfo[i] = inputShapeInfo[i];

			shape::updateStrides(outShapeInfo, shape::order(inputShapeInfo));

    		return SHAPELIST(outShapeInfo);
		}


}
}

#endif